MapleServer2/Types/FieldNavigator.cs

```diff
@@ -0,0 +1,244 @@
+    public List<CoordS> FindPath(Agent agent, CoordS endCoord)
+    {
+        Position end = FindPositionFromCoordS(endCoord);
+
+        if (!Mesh.positionIsValid(end))
+        {
+            return null;
+        }
+
+        using Path shortestPath = agent.findShortestPathTo(CollisionContext, end);
+        if (shortestPath is null)
+        {
+            Console.WriteLine("Shortest path is null");
+            return null;
+        }
+
+        using Path path = agent.generateCurvedPath(shortestPath, CollisionContext, 0, 0, sectionLength: 50, turnRatio1: 0.5f, turnRatio2: 0.9f);
+        if (path is null)
+        {
+            Console.WriteLine("Curved path is null");
+        }
+
+        return PathToCoordS(path);
+    }
+
+    /// <summary>
+    /// Creates a box from the given capsule metadata. Used to calculate collision in the mesh.
+    /// </summary>
+    public Shape AddShape(NpcMetadataCapsule metadata)
+    {
+        // TODO: Cache shapes??
+        int width = metadata.Radius; // Using radius for width
+        int height = metadata.Height;
+
+        int halfWidth = width / 2;
+        int halfHeight = height / 2;
+        int negativeHalfWidth = halfWidth * -1;
+        int negativeHalfHeight = halfHeight * -1;
+        List<Point> rectArray = new()
+        {
+            new(negativeHalfWidth, negativeHalfHeight),
+            new(negativeHalfWidth, halfHeight),
+            new(halfWidth, halfHeight),
+            new(halfWidth, negativeHalfHeight)
+        };
+
+        Shape shape = MapleServer.PathEngine.newShape(rectArray);
+        Mesh.generateUnobstructedSpaceFor(shape, true);
+        Mesh.generatePathfindPreprocessFor(shape);
+        return shape;
+    }
+
+    /// <summary>
+    /// Creates an agent from the given sh2ape, also adds it to the mesh and to the collision context.
+    /// </summary>
+    /// <returns>Agent or null if position isn't valid</returns>
+    public Agent AddAgent(IFieldActor actor, Shape shape)
+    {
+        Position position = FindPositionFromCoordS(actor.Coord.ToShort());
+        if (!Mesh.positionIsValid(position))
+        {
+            return null;
+        }
+
+        Agent agent = Mesh.placeAgent(shape, position);
+        agent.setUserData(actor.ObjectId);
+
+        CollisionContext.addAgent(agent);
+        return agent;
+    }
+
+    /// <summary>
+    /// Find the nearest position on the mesh to the given coordS.
+    /// </summary>
+    public Position FindPositionFromCoordS(CoordS coordF)
+    {
+        return Mesh.positionNear3DPoint(coordF.X, coordF.Y, coordF.Z, horizontalRange: 15, verticalRange: 15);
+    }
+
+    /// <summary>
+    /// Check if the given position is valid.
+    /// </summary>
+    public bool PositionIsValid(Position position)
+    {
+        return Mesh.positionIsValid(position);
+    }
+
+    public Position? FindClosestUnobstructedPosition(Shape shape, Position position, int radius)
```

Comment:
rather than all these `PositionIsValid` checks, you should be able to wrap the code with try/catch `InvalidPositionException`.